zero/gkyl_position_map_priv.h

```diff
@@ -0,0 +1,448 @@
+    for (int iz = 0; iz < scan_cells; iz++)
+    {
+      double left_xi = scan_left + iz * scan_dxi;
+      double right_xi = scan_left + (iz + 1) * scan_dxi;
+      double psi = constB_ctx->psi;
+      double alpha = constB_ctx->alpha;
+
+      double left_theta[1], right_theta[1];
+      position_map_constB_z_polynomial(0.0, &left_xi, left_theta, constB_ctx);
+      position_map_constB_z_polynomial(0.0, &right_xi, right_theta, constB_ctx);
+
+      xp[Z_IDX] = left_theta[0];
+      gkyl_calc_bmag_global(0.0, xp, fout, bmag_ctx);
+      double Bmag_left = fout[0];
+      xp[Z_IDX] = right_theta[0];
+      gkyl_calc_bmag_global(0.0, xp, fout, bmag_ctx);
+      double Bmag_right = fout[0];
+      double dB_dCell = (Bmag_right - Bmag_left);
+      if (fabs(dB_dCell) > max_dB_dCell)
+      {
+        max_dB_dCell = fabs(dB_dCell);
+      }
+    }
+    double improvement = max_dB_dCell_prior - max_dB_dCell;
+    if (improvement > 1e-3)
+    {
+      expander_order++;
+      max_dB_dCell_prior = max_dB_dCell;
+    }
+    else if (improvement < 0)
+    {
+      expander_order--;
+      constB_ctx->map_order_expander = expander_order;
+      break;
+    }
+    else
+    {
+      break;
+    }
+
+  }
+  double max_dB_dCell_expander = max_dB_dCell;
+  //Center region
+  scan_left = 0.0;
+  scan_right = constB_ctx->theta_throat;
+  scan_dxi = (scan_right - scan_left) / scan_cells;
+  int center_order = 1;
+  max_dB_dCell_prior = 99999999.99;
+  while (1)
+  {
+    max_dB_dCell = 0.0;
+    constB_ctx->map_order_center = center_order;
+    for (int iz = 0; iz < scan_cells; iz++)
+    {
+      double left_xi = scan_left + iz * scan_dxi;
+      double right_xi = scan_left + (iz + 1) * scan_dxi;
+
+      double left_theta[1], right_theta[1];
+      position_map_constB_z_polynomial(0.0, &left_xi, left_theta, constB_ctx);
+      position_map_constB_z_polynomial(0.0, &right_xi, right_theta, constB_ctx);
+
+      xp[Z_IDX] = left_theta[0];
+      gkyl_calc_bmag_global(0.0, xp, fout, bmag_ctx);
+      double Bmag_left = fout[0];
+      xp[Z_IDX] = right_theta[0];
+      gkyl_calc_bmag_global(0.0, xp, fout, bmag_ctx);
+      double Bmag_right = fout[0];
+
+      double dB_dCell = (Bmag_right - Bmag_left);
+      if (fabs(dB_dCell) > max_dB_dCell)
+      {
+        max_dB_dCell = fabs(dB_dCell);
+      }
+    }
+    double improvement = max_dB_dCell_prior - max_dB_dCell;
+    if (improvement > 1e-3)
+    {
+      center_order++;
+      max_dB_dCell_prior = max_dB_dCell;
+    }
+    else if (improvement < 0)
+    {
+      center_order--;
+      constB_ctx->map_order_center = center_order;
+      break;
+    }
+    else
+    {
+      break;
+    }
+  }
+  free(xp);
+  free(fout);
+}
+
+
+// Utility functions for numeric root finding B mapping
+
+
+double
+calc_bmag_global_derivative(double xn, void *ctx)
+{
+  struct gkyl_position_map *gpm = ctx;
+  struct gkyl_bmag_ctx *bmag_ctx = gpm->bmag_ctx;
+  double h = 1e-6;
```

Comment:
why 1e-6? should this be multiplied by something else? what if my field line is 2e-6 long?